map-downloader-android/app/src/main/java/com/here/android/example/map/downloader/MainActivity.java

```diff
@@ -18,48 +18,55 @@
 
 import android.Manifest;
 import android.app.ListActivity;
+import android.content.Context;
 import android.content.pm.PackageManager;
+import android.os.Build;
 import android.os.Bundle;
 import android.support.annotation.NonNull;
 import android.support.v4.app.ActivityCompat;
 import android.view.View;
 import android.widget.ListView;
 import android.widget.Toast;
 
-import java.util.ArrayList;
-import java.util.List;
-
 /**
  * Main activity which launches the map list view and handles Android run-time requesting
  * permission.
  */
-
 public class MainActivity extends ListActivity {
-    private final static int REQUEST_CODE_ASK_PERMISSIONS = 1;
+    private static final int REQUEST_CODE_ASK_PERMISSIONS = 1;
+    private static final String[] RUNTIME_PERMISSIONS = {
+            Manifest.permission.ACCESS_FINE_LOCATION,
+            Manifest.permission.WRITE_EXTERNAL_STORAGE,
+            Manifest.permission.INTERNET,
+            Manifest.permission.ACCESS_WIFI_STATE,
+            Manifest.permission.ACCESS_NETWORK_STATE};
     private MapListView m_mapListView;
 
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         setContentView(R.layout.activity_main);
-        requestPermissions();
+
+        if (hasPermissions(this, RUNTIME_PERMISSIONS)) {
+            setupMapListView();
+        } else {
+            ActivityCompat.requestPermissions(this, RUNTIME_PERMISSIONS, REQUEST_CODE_ASK_PERMISSIONS);
+        }
     }
 
     /**
      * Only when the app's target SDK is 23 or higher, it requests each dangerous permissions it
      * needs when the app is running.
      */
-    private void requestPermissions() {
-        final List<String> requiredSDKPermissions = new ArrayList<>();
-        requiredSDKPermissions.add(Manifest.permission.ACCESS_FINE_LOCATION);
-        requiredSDKPermissions.add(Manifest.permission.WRITE_EXTERNAL_STORAGE);
-        requiredSDKPermissions.add(Manifest.permission.INTERNET);
-        requiredSDKPermissions.add(Manifest.permission.ACCESS_WIFI_STATE);
-        requiredSDKPermissions.add(Manifest.permission.ACCESS_NETWORK_STATE);
-
-        ActivityCompat.requestPermissions(this,
-                requiredSDKPermissions.toArray(new String[requiredSDKPermissions.size()]),
-                REQUEST_CODE_ASK_PERMISSIONS);
+    private static boolean hasPermissions(Context context, String... permissions) {
+        if (android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && context != null && permissions != null) {
```

Comment:
Asking for *no* permissions should yield true. I'll remove the context check and rather raise because true / false would be both wrong in case of `context==null`